### PR TITLE
Create mirror pod before running a static pod so that the API can be informed before calling plugins.

### PR DESCRIPTION
**Pull request by alexhersh**
I have a potential fix for #14992.
I have yet to prepare this PR for merging, but I was hoping to get some feedback on the design before I do so.

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1458,6 +1458,13 @@ func (kl *Kubelet) syncPod(pod *api.Pod, mirrorPod *api.Pod, runningPod kubecont
 		return err
 	}
 
+	if kubepod.IsStaticPod(pod) && mirrorPod == nil {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Uh..this does not prevent line 1576 to create the mirror pod again.

**Comment by yujuhong** [comment]
How do you guarantee the mirror pod has been created before syncing the pod. We don't fail the syncPod function if there is a mirror pod creation error. Also, we don't want to fail the function because this has to work with standalone kubelet too.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (alexhersh) [inline pkg/kubelet/kubelet.go]: Forgot to update `mirrorPod` with the fresh one. Now the block at 1567 will update the mirror.

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1458,6 +1458,15 @@ func (kl *Kubelet) syncPod(pod *api.Pod, mirrorPod *api.Pod, runningPod kubecont
...
 
+	// Create Mirror Pod for Static Pod if it doesn't already exist
+	if kubepod.IsStaticPod(pod) && mirrorPod == nil {
+		glog.V(3).Infof("Creating a mirror pod for static pod %q", podFullName)
+		if err := kl.podManager.CreateMirrorPod(pod); err != nil {
+			glog.Errorf("Failed creating a mirror pod %q: %v", podFullName, err)
+		}
+		mirrorPod, _ = kl.podManager.GetMirrorPodByPod(pod)
+	}
+
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Sorry that I didn't express more clear in the previous comment. What's the point of creating the pod at the end. I think you can remove the entire block.

Comment (alexhersh) [inline pkg/kubelet/kubelet.go]: My thought was that pod status could change during sync, and thus the mirror needed to be updated, but on second thought that may not be necessary.

Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Creating a pod is different than updating the status. We only need to pod spec to create the mirror pod, and pod spec should not change during the sync. I think it's fine to remove them.

**Comment by alexhersh** [comment]
@yujuhong, do you have any recommendations on how to ensure that the mirrorpod is finished before we run sync?

**Comment by yujuhong** [comment]
> @yujuhong, do you have any recommendations on how to ensure that the mirrorpod is finished before we run sync?

@alexhersh, would it be okay to simply retry the next sync or would it be undesirable?

**Comment by caseydavenport** [comment]
@yujuhong - I think that is probably the right approach.  Pod creation might fail the first time around, but by the next sync the mirror pod will exist and all will be fine.

**Comment by yujuhong** [comment]
LGTM. Will apply the label after the commits are squashed.

**Comment by alexhersh** [comment]
squashed!

**Comment by alexhersh** [comment]
My e2e test failed but I can't see the results

**Comment by yujuhong** [comment]
@alexhersh, it's a test infra problem. See #17257

**Comment by piosz** [comment]
@k8s-bot e2e test this please

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Comment by wojtek-t** [comment]
@k8s-bot e2e test this please

**Comment by caseydavenport** [comment]
@yujuhong - Any idea when this fix will be included in a release?  Will it be in the next v1.1.X release?

**Comment by yujuhong** [comment]
@caseydavenport, we can cherry-pick this to the 1.1 branch for the next release. @alexhersh, would you like to create a PR?

**Comment by alexhersh** [comment]
@yujuhong  I made #17597 against the release-1.1 branch

**Pull request merged**